### PR TITLE
Replace comparator with java8 way

### PR DESCRIPTION
**Pull request by liketic**
This PR is used to upgrade ```Comparator``` with lambda of java 8 which will make the code much shorter and cleaner. Happy to hear your comments. Thanks.

### CODE REVIEW

**Comment by jpountz** [comment]
An alternative would be to use method references instead, ie. `Comparator.comparingInt(ActionFilter::order)` rather than `(o1, o2) -> Integer.compare(o1.order(), o2.order())`. I don't have strong feelings about it, maybe a slight preference for method references. I'll wait and see what others think before merging.

**Comment by jasontedor** [comment]
I was about to comment the same thing as @jpountz, I prefer the method references where possible. Also, I see at least one change that is not a comparator change (in HppcMaps); can we revert that change and any other such changes and just keep this focused on the comparators for now?

File: core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java [hunk 1]
```diff
@@ -256,7 +256,7 @@ public synchronized boolean isEnoughPendingJoins(int pendingMasterJoins) {
 
         private Map<DiscoveryNode, ClusterStateTaskListener> getPendingAsTasks() {
             Map<DiscoveryNode, ClusterStateTaskListener> tasks = new HashMap<>();
-            joinRequestAccumulator.entrySet().stream().forEach(e -> tasks.put(e.getKey(), new JoinTaskListener(e.getValue(), logger)));
+            joinRequestAccumulator.entrySet().forEach(e -> tasks.put(e.getKey(), new JoinTaskListener(e.getValue(), logger)));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/common/network/NetworkUtils.java [hunk 2]
```diff
@@ -208,16 +199,14 @@ public static boolean defaultReuseAddress() {
...
+    /** Returns all addresses (any scope) for interfaces that are up.
      *  This is only used to pick a publish address, when the user set network.host to a wildcard */
     static InetAddress[] getAllAddresses() throws SocketException {
         List<InetAddress> list = new ArrayList<>();
         for (NetworkInterface intf : getInterfaces()) {
             if (intf.isUp()) {
-                for (InetAddress address : Collections.list(intf.getInetAddresses())) {
-                    list.add(address);
-                }
+                list.addAll(Collections.list(intf.getInetAddresses()));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/network/NetworkUtils.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/common/collect/HppcMaps.java [hunk 3]
```diff
@@ -122,12 +122,7 @@ public void remove() {
                 throw new UnsupportedOperationException();
             }
         };
-        return new Iterable<T>() {
-            @Override
-            public Iterator<T> iterator() {
-                return intersection;
-            }
-        };
+        return () -> intersection;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/collect/HppcMaps.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 4]
```diff
@@ -91,7 +90,7 @@
             Property.Dynamic, Property.NodeScope);
     @Deprecated
     public static final Setting<String> INDEX_RECOVERY_INITIAL_SHARDS_SETTING =
-        new Setting<>("index.recovery.initial_shards", (settings) -> NODE_INITIAL_SHARDS_SETTING.get(settings) , INITIAL_SHARDS_PARSER,
+        new Setting<>("index.recovery.initial_shards", NODE_INITIAL_SHARDS_SETTING::get, INITIAL_SHARDS_PARSER,
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 5]
```diff
@@ -509,7 +509,7 @@ private void setFieldValue(String path, Object value, boolean append) {
     private static void appendValues(List<Object> list, Object value) {
         if (value instanceof List) {
             List<?> valueList = (List<?>) value;
-            valueList.stream().forEach(list::add);
+            valueList.forEach(list::add);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 6]
```diff
@@ -337,15 +329,12 @@ public TopDocs call() throws Exception {
...
                     collector = null;
-                    topDocsCallable = new Callable<TopDocs>() {
-                        @Override
-                        public TopDocs call() throws Exception {
-                            int count = 0;
-                            for (LeafReaderContext context : searcher.getIndexReader().leaves()) {
-                                count += context.reader().docFreq(term);
-                            }
-                            return new TopDocs(count, Lucene.EMPTY_SCORE_DOCS, 0);
+                    topDocsCallable = () -> {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 7]
```diff
@@ -163,12 +163,7 @@ static boolean execute(SearchContext searchContext, final IndexSearcher searcher
                 if (searchContext.getProfilers() != null) {
                     collector = new InternalProfileCollector(collector, CollectorResult.REASON_SEARCH_COUNT, Collections.emptyList());
                 }
-                topDocsCallable = new Callable<TopDocs>() {
-                    @Override
-                    public TopDocs call() throws Exception {
-                        return new TopDocs(totalHitCountCollector.getTotalHits(), Lucene.EMPTY_SCORE_DOCS, 0);
-                    }
-                };
+                topDocsCallable = () -> new TopDocs(totalHitCountCollector.getTotalHits(), Lucene.EMPTY_SCORE_DOCS, 0);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 8]
```diff
@@ -222,36 +217,33 @@ public TopDocs call() throws Exception {
...
-                            case DFS_QUERY_AND_FETCH:
-                                // for (DFS_)QUERY_AND_FETCH, we already know the last emitted doc
-                                if (topDocs.scoreDocs.length > 0) {
-                                    // set the last emitted doc
-                                    scrollContext.lastEmittedDoc = topDocs.scoreDocs[topDocs.scoreDocs.length - 1];
-                                }
-                                break;
-                            default:
-                                break;
+                topDocsCallable = () -> {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java [hunk 9]
```diff
@@ -329,7 +330,7 @@ public static FieldSortBuilder fromXContent(QueryParseContext context, String fi
     private static ObjectParser<FieldSortBuilder, QueryParseContext> PARSER = new ObjectParser<>(NAME);
 
     static {
-        PARSER.declareField(FieldSortBuilder::missing, p -> p.objectText(),  MISSING, ValueType.VALUE);
+        PARSER.declareField(FieldSortBuilder::missing, XContentParser::objectText,  MISSING, ValueType.VALUE);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java]: Please revert this change.

File: core/src/test/java/org/elasticsearch/rest/RestFilterChainTests.java [hunk 10]
```diff
@@ -124,12 +113,7 @@ public void testTooManyContinueProcessing() throws Exception {
         RestController restController = new RestController(Settings.EMPTY, Collections.emptySet());
         restController.registerFilter(testFilter);
 
-        restController.registerHandler(RestRequest.Method.GET, "/", new RestHandler() {
-            @Override
-            public void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
-                channel.sendResponse(new TestResponse());
-            }
-        });
+        restController.registerHandler(RestRequest.Method.GET, "/", (request, channel, client) -> channel.sendResponse(new TestResponse()));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/rest/RestFilterChainTests.java]: Please revert this change.

**Review by jasontedor — CHANGES_REQUESTED** [review]
To keep the surface area of the change smaller, can you please revert the changes that are unrelated to the comparator change which appears to be the main point of this pull request?

File: core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java]: None of these changes are necessary, but if we are going to get in these places and make changes, let's do it the preferred way which is with method references (where possible).

**Pull request closed without merging**